File: script/duck_detection.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image, CameraInfo
from cv_bridge import CvBridge

import cv2
import numpy as np
from openvino import Core


MODEL_PATH = "last_int8_openvino_model/last.xml"
IMG_SIZE = 320
CONF_THRES = 0.4


class OpenVINO3DNode(Node):

    def __init__(self):
        super().__init__("openvino_3d_node")

        self.bridge = CvBridge()
        self.depth_image = None

        self.fx = None
        self.fy = None
        self.cx0 = None
        self.cy0 = None

        # ===== OpenVINO =====
        core = Core()
        model = core.read_model(MODEL_PATH)
        self.compiled_model = core.compile_model(model, "CPU")
        self.output_layer = self.compiled_model.output(0)

        # ===== Publisher =====
        self.image_pub = self.create_publisher(
            Image,
            "/detection/duck_bbox_image",
            10)

        # ===== Subscriptions =====
        self.create_subscription(
            Image,
            "/camera/camera/color/image_raw",
            self.color_callback,
            10)

        self.create_subscription(
            Image,
            "/camera/camera/aligned_depth_to_color/image_raw",
            self.depth_callback,
            10)

        self.create_subscription(
            CameraInfo,
            "/camera/camera/color/camera_info",
            self.camera_info_callback,
            10)

        self.get_logger().info("OpenVINO 3D Node (Publish Mode) Started")

    # =====================================
    def camera_info_callback(self, msg):
        self.fx = msg.k[0]
        self.fy = msg.k[4]
        self.cx0 = msg.k[2]
        self.cy0 = msg.k[5]

    # =====================================
    def depth_callback(self, msg):
        self.depth_image = self.bridge.imgmsg_to_cv2(
            msg, desired_encoding="passthrough")

    # =====================================
    def color_callback(self, msg):

        if self.depth_image is None or self.fx is None:
            return

        frame = self.bridge.imgmsg_to_cv2(msg, "bgr8")
        H, W = frame.shape[:2]

        # ===== 前処理 =====
        img = cv2.resize(frame, (IMG_SIZE, IMG_SIZE))
        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        img = img.astype(np.float32) / 255.0
        img = np.transpose(img, (2, 0, 1))
        img = np.expand_dims(img, axis=0)

        # ===== 推論 =====
        results = self.compiled_model([img])[self.output_layer]
        detections = results[0]

        scale_x = W / IMG_SIZE
        scale_y = H / IMG_SIZE

        for det in detections:
            x1, y1, x2, y2, score, cls = det

            if score < CONF_THRES:
                continue

            x1 = int(x1 * scale_x)
            y1 = int(y1 * scale_y)
            x2 = int(x2 * scale_x)
            y2 = int(y2 * scale_y)

            u = int((x1 + x2) / 2)
            v = int(y1 + (y2 - y1) * 0.6)

            # ===== depth 安定取得 =====
            h, w = self.depth_image.shape
            r = 2

            u_min = max(u - r, 0)
            u_max = min(u + r + 1, w)
            v_min = max(v - r, 0)
            v_max = min(v + r + 1, h)

            depth_roi = self.depth_image[v_min:v_max, u_min:u_max]
            valid_depth = depth_roi[depth_roi > 0]

            if len(valid_depth) < 1:
                continue

            depth = np.median(valid_depth)
            z = float(depth) / 1000.0

            if z <= 0:
                continue

            # ===== 3D計算 =====
            X = (u - self.cx0) * z / self.fx
            Y = (v - self.cy0) * z / self.fy
            Z = z

            # ===== 描画 =====
            cv2.rectangle(frame, (x1, y1), (x2, y2), (0, 255, 0), 2)

            text = f"X:{X:.2f} Y:{Y:.2f} Z:{Z:.2f}"
            cv2.putText(frame,
                        text,
                        (x1, y1 - 10),
                        cv2.FONT_HERSHEY_SIMPLEX,
                        0.6,
                        (0, 255, 0),
                        2)

        # ===== Publish =====
        out_msg = self.bridge.cv2_to_imgmsg(frame, encoding="bgr8")
        out_msg.header = msg.header  # タイムスタンプ引き継ぎ
        self.image_pub.publish(out_msg)


def main(args=None):
    rclpy.init(args=args)
    node = OpenVINO3DNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/detection/krb2026b_duck_detection/src/duck_detection.cpp
#include "krb2026b_duck_detection/duck_detection.hpp"

namespace duck_detection {
duck_detection::duck_detection (const rclcpp::NodeOptions &options) : Node ("duck_detection", options), cam_info_ready_ (false) {
    this->declare_parameter<std::string> ("model_path", "last_int8_openvino_model/last.xml");

    std::string model_path = this->get_parameter ("model_path").as_string ();
    ov::Core    core;
    auto        model = core.read_model (model_path);
    compiled_model_   = core.compile_model (model, "CPU");
    output_layer_     = compiled_model_.output (0);

    image_pub_ = create_publisher<sensor_msgs::msg::Image> ("duck_bbox_image", 10);
    point_pub_ = create_publisher<geometry_msgs::msg::PointStamped> ("duck_position", 10);

    color_sub_   = create_subscription<sensor_msgs::msg::Image> ("color_image_raw", 10, std::bind (&duck_detection::colorCallback, this, std::placeholders::_1));
    depth_sub_   = create_subscription<sensor_msgs::msg::Image> ("aligned_depth_to_color_image_raw", 10, std::bind (&duck_detection::depthCallback, this, std::placeholders::_1));
    caminfo_sub_ = create_subscription<sensor_msgs::msg::CameraInfo> ("camera_info", 10, std::bind (&duck_detection::cameraInfoCallback, this, std::placeholders::_1));
    tf_buffer_   = std::make_unique<tf2_ros::Buffer> (this->get_clock ());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener> (*tf_buffer_);

    RCLCPP_INFO (get_logger (), "duck detection Node Started");
}
void duck_detection::colorCallback (const sensor_msgs::msg::Image::SharedPtr msg) {
    if (depth_image_.empty () || !cam_info_ready_) return;

    auto    cv_ptr = cv_bridge::toCvCopy (msg, "bgr8");
    cv::Mat frame  = cv_ptr->image;

    int H = frame.rows;
    int W = frame.cols;

    // ===== Preprocess =====
    cv::Mat resized;
    cv::resize (frame, resized, cv::Size (IMG_SIZE, IMG_SIZE));
    cv::cvtColor (resized, resized, cv::COLOR_BGR2RGB);
    resized.convertTo (resized, CV_32F, 1.0 / 255.0);

    // HWC → CHW
    std::vector<float> input_data (3 * IMG_SIZE * IMG_SIZE);

    for (int c = 0; c < 3; ++c) {
        for (int y = 0; y < IMG_SIZE; ++y) {
            for (int x = 0; x < IMG_SIZE; ++x) {
                input_data[c * IMG_SIZE * IMG_SIZE + y * IMG_SIZE + x] = resized.at<cv::Vec3f> (y, x)[c];
            }
        }
    }

    ov::Tensor input_tensor (ov::element::f32, {1, 3, IMG_SIZE, IMG_SIZE}, input_data.data ());

    auto infer_request = compiled_model_.create_infer_request ();
    infer_request.set_input_tensor (input_tensor);
    infer_request.infer ();

    // ===== Postprocess =====
    auto output = infer_request.get_output_tensor ();
    auto shape  = output.get_shape ();  // [1, 2100, 6]

    float *data = output.data<float> ();

    int num_boxes = shape[1];  // 2100
    int elements  = shape[2];  // 6

    float scale_x = static_cast<float> (W) / IMG_SIZE;
    float scale_y = static_cast<float> (H) / IMG_SIZE;

    double use_x = 0.0;
    double use_y = 0.0;
    double use_z = 0.0;

    for (int i = 0; i < num_boxes; ++i) {
        float x1    = data[i * elements + 0];
        float y1    = data[i * elements + 1];
        float x2    = data[i * elements + 2];
        float y2    = data[i * elements + 3];
        float score = data[i * elements + 4];
        float cls   = data[i * elements + 5];

        if (score < CONF_THRES) continue;

        x1 *= scale_x;
        y1 *= scale_y;
        x2 *= scale_x;
        y2 *= scale_y;

        // ===== 中心点計算 =====
        int u = static_cast<int> ((x1 + x2) / 2.0);
        int v = static_cast<int> (y1 + (y2 - y1) * 0.6);

        // ===== depth 安定取得 =====
        int h = depth_image_.rows;
        int w = depth_image_.cols;
        int r = 2;

        int u_min = std::max (u - r, 0);
        int u_max = std::min (u + r + 1, w);
        int v_min = std::max (v - r, 0);
        int v_max = std::min (v + r + 1, h);

        if (u_min >= u_max || v_min >= v_max) continue;

        // depthは通常16UC1（mm）
        cv::Mat depth_roi = depth_image_ (cv::Range (v_min, v_max), cv::Range (u_min, u_max));

        std::vector<double> valid_depths;

        for (int yy = 0; yy < depth_roi.rows; ++yy) {
            for (int xx = 0; xx < depth_roi.cols; ++xx) {
                uint16_t d = depth_roi.at<uint16_t> (yy, xx);

                if (d > 0) valid_depths.push_back (static_cast<double> (d));
            }
        }

        if (valid_depths.empty ()) continue;

        // median計算
        std::nth_element (valid_depths.begin (), valid_depths.begin () + valid_depths.size () / 2, valid_depths.end ());

        double depth = valid_depths[valid_depths.size () / 2];

        if (depth <= 0.0) continue;

        // ===== 3D計算 =====
        double z = depth / 1000.0;  // mm → m
        double x = (u - cx0_) * z / fx_;
        double y = (v - cy0_) * z / fy_;
        if (use_x == 0 || x > use_x) {
            use_x = x;
            use_y = y;
            use_z = z;
        }
        // ===== 描画 =====
        cv::rectangle (frame, cv::Point (x1, y1), cv::Point (x2, y2), cv::Scalar (0, 255, 0), 2);

        std::ostringstream oss;
        oss << std::fixed << std::setprecision (2) << "X:" << x << " Y:" << y << " Z:" << z;

        cv::putText (frame, oss.str (), cv::Point (x1, y1 - 10), cv::FONT_HERSHEY_SIMPLEX, 0.6, cv::Scalar (0, 255, 0), 2);
    }

    auto out_msg = cv_bridge::CvImage (msg->header, "bgr8", frame).toImageMsg ();
    image_pub_->publish (*out_msg);

    if (use_x == 0.0 && use_y == 0.0 && use_z == 0.0) return;
    if (use_z > 5.0) return;

    geometry_msgs::msg::PointStamped point_msg;
    point_msg.header  = msg->header;
    point_msg.point.x = use_x;
    point_msg.point.y = use_y;
    point_msg.point.z = use_z;
    try {
        map_point = tf_buffer_->transform (point_msg, "map", tf2::durationFromSec (0.1));

    } catch (tf2::TransformException &ex) {
        RCLCPP_WARN (this->get_logger (), "%s", ex.what ());
        return;
    }
    if (map_point.point.x < 3.0 || map_point.point.x > 4.0) return;
    point_pub_->publish (map_point);
}
void duck_detection::cameraInfoCallback (const sensor_msgs::msg::CameraInfo::SharedPtr msg) {
    fx_             = msg->k[0];
    fy_             = msg->k[4];
    cx0_            = msg->k[2];
    cy0_            = msg->k[5];
    cam_info_ready_ = true;
}

void duck_detection::depthCallback (const sensor_msgs::msg::Image::SharedPtr msg) {
    depth_image_ = cv_bridge::toCvCopy (msg, msg->encoding)->image;
}
}  // namespace duck_detection

#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE (duck_detection::duck_detection)